solver/src/solver/paraswap_solver.rs

```diff
@@ -111,43 +169,47 @@ impl SingleOrderSolving for ParaswapSolver {
                     / 10000,
             ),
         };
-        let transaction_query = TransactionBuilderQuery {
+        let query = TransactionBuilderQuery {
             src_token: order.sell_token,
             dest_token: order.buy_token,
             src_amount,
             dest_amount,
-            from_decimals: decimals(&order.sell_token)?,
-            to_decimals: decimals(&order.buy_token)?,
-            price_route: price_response.price_route_raw,
+            from_decimals: decimals(token_info, &order.sell_token)?,
+            to_decimals: decimals(token_info, &order.buy_token)?,
+            price_route: price_response.clone().price_route_raw,
             user_address: self.solver_address,
             referrer: REFERRER.to_string(),
         };
-        let transaction = self.client.transaction(transaction_query).await?;
-
-        let mut settlement = Settlement::new(hashmap! {
-            order.sell_token => price_response.dest_amount,
-            order.buy_token => price_response.src_amount,
-        });
-        settlement.with_liquidity(&order, amount)?;
-
-        settlement.encoder.append_to_execution_plan(
-            self.allowance_fetcher
-                .get_approval(
-                    order.sell_token,
-                    APPROVAL_RECEIVER,
-                    price_response.src_amount,
-                )
-                .await?,
-        );
-        settlement.encoder.append_to_execution_plan(transaction);
-        Ok(Some(settlement))
+        Ok(query)
     }
 
-    fn name(&self) -> &'static str {
-        "ParaSwap"
+    async fn paraswap_transaction_from_order(
+        &self,
+        order: &LimitOrder,
+    ) -> Result<ParaswapCompleteResponse> {
+        let token_info = self
+            .token_info
+            .get_token_infos(&[order.sell_token, order.buy_token])
+            .await;
+        let (price_response, amount) = self.get_price_for_order(&order, &token_info).await?;
+        let transaction_query = self
```

Comment:
This is a good point and something that I didn't consider.

Reply:
I am not entirely sure what is expected as a proposed change to this PR. Does this mean some of our Error classes are no longer relevant?

Reply:
I think the suggestion is to check the limit price before building the transaction.

Reply:
I have commented on this here

https://github.com/gnosis/gp-v2-services/pull/812#discussion_r668112376

The problem with this that we have now a bunch of helper methods that combine these queries and the return type of checking the limit price after price query is a `None` settlement. which doesn't really fall under the category of the combined method return types... at this moment (in the current PR) we are maximally inefficient about returning None when limit price isn't satisfied because we make both price and transaction queries as many times as necessary (given the retry limit) and only after successful build do we check the limit price.

I can imagine reworking some of the error types yet again to account for this, but given the current implementation don't see how we can "return early with None"

Reply:
You can move the limit price check from `settle_order` to `paraswap_transaction_from_order` in between getting the price quote and building the transaction. In pseudo code:
```
fn settle_order() {
  for 0..MAX_RETRIES {
    match try_settle_order() {
      Ok(settlement) => return Ok(settlement),
      Err(err) if is_retryable_err(err) => continue,
      Err(err) => return Err(err.into()),
  }
}

fn try_settle_order() {
  let quote = api.get_price()?;
  if !is_limit_price_satisfied(quote, order) {
    return Ok(None);
  }
  let tx = api.build_transaction()?;
  return Ok(...);
}
```

Reply:
I will rework this in to something similar to your suggestion.

Reply:
The issue I have with this suggestion (as the reason I didn't do it this way before) is that there are many other `anyhow` errors that are not `ParaswapResponseErrors` which can occur during your suggested `try_settle_order`. If there is some way to retain the ParaswapErrors while still dropping a bunch of ?s all over to compensate for all the other possible errors then this would surely be the way to go...

Reply:
Ok, I have since overcome this issue by introducing a new error enum 

SettlementError {
  Anyhow(anyhow::Error),
  Paraswap(ParaswapResponseError),
}

along with a few conversion methods.

It is now structured much simpler and as suggested in the pseudo code. I hope this is looking more like what we are going for. 

Will try and introduce some tests tomorrow morning!